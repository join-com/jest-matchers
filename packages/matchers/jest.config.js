module.exports = {
  testEnvironment: 'node',
  testPathIgnorePatterns: ['/node_modules/'],
  transform: { '^.+\\.tsx?$': ['ts-jest', { diagnostics: false }] },
  maxWorkers: 1,
  workerIdleMemoryLimit: '1.5GB',
}
