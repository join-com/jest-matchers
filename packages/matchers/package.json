{
  "name": "@join-private/jest-matchers",
  "version": "0.0.0-development",
  "description": "Set of custom jest matchers",
  "author": "JOIN Solutions AG",
  "license": "UNLICENSED",
  "main": "lib/index.js",
  "types": "types/index.d.ts",
  "files": [
    "lib",
    "types/index.d.ts"
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/join-com/jest-matchers"
  },
  "scripts": {
    "build": "tsc -p tsconfig.prod.json",
    "test": "jest",
    "test:ci": "jest --no-color",
    "test:watch": "jest --watch",
    "lint": "yarn lint:tsc && yarn lint:eslint",
    "lint:eslint": "eslint . --ext .ts --max-warnings 0",
    "lint:tsc": "tsc --noEmit",
    "prepublishOnly": "yarn lint && yarn build"
  },
  "keywords": [
    "jest",
    "matchers"
  ],
  "engines": {
    "node": ">=14.0.0"
  },
  "devDependencies": {
    "@join-private/eslint-config-backend": "^1.5.0",
    "@types/expect": "^24.3.0",
    "@types/jest": "^29.5.12",
    "@types/node": "^22.5.4",
    "eslint": "^8.18.0",
    "jest": "^29.7.0",
    "prettier": "^3.3.3",
    "ts-jest": "^29.2.5",
    "typescript": "^5.5.4"
  },
  "dependencies": {
    "expect": "^29.7.0",
    "fast-deep-equal": "^3.1.3",
    "jest-matcher-utils": "29.7.0"
  },
  "publishConfig": {
    "access": "restricted",
    "registry": "https://europe-west1-npm.pkg.dev/join-dev-330413/join-private/"
  }
}
