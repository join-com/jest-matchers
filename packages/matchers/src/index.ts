/* eslint-disable no-console */
/* eslint-disable @typescript-eslint/no-unsafe-call */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
import * as matchers from './matchers'

const jestExpect = global['expect']

if (jestExpect !== undefined) {
  jestExpect.extend(matchers)
} else {
  console.error(`
    Unable to find Jest's global expect.
    Please check you have added jest-matchers to setupFilesAfterEnv section of jest config
  `)
}
