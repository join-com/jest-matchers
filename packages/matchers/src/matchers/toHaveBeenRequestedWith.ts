import { diff, matcherHint, printExpected, printReceived, RECEIVED_COLOR } from 'jest-matcher-utils'

type ICallWithRequest = [{ request: Record<string, unknown> }]

interface IMock {
  mock: {
    calls: ICallWithRequest[]
  }
}

export const toHaveBeenRequestedWith = ({ mock }: IMock, expected: unknown): jest.CustomMatcherResult => {
  const calls = mock.calls
  const pass = calls.length > 0 && anyOfRequestsSucceeded(calls, expected)
  return {
    pass,
    message: () => {
      const hintString = matcherHint('.toHaveBeenRequestedWith')
      const expectedString = printExpected(expected)
      const mismatchString = formatMismatchedCalls(mock.calls, expected)
      return `${hintString}\n\n` + `Expected request value to equal:\n${expectedString}\n` + mismatchString
    },
  }
}

const formatMismatchedCalls = (calls: ICallWithRequest[], expected: unknown): string => {
  if (!calls.length) {
    return `But it was ${RECEIVED_COLOR('not called')}.`
  }

  const messages: string[] = ['Received:']
  calls.map((call, index) => {
    const received = call[0].request
    const diffString = diff(expected, received) ?? ''

    const receivedString = printReceived(received)
    const callNumber: string = calls.length > 1 ? `\ncall no.${String(index + 1)}\n` : ''
    const callMessage = `${callNumber}` + `${receivedString}\n` + 'Difference:\n' + `${diffString}`

    messages.push(callMessage)
  })

  if (calls.length > 1) {
    messages.push(`\nNumber of calls: ${calls.length}`)
  }
  return messages.join('\n')
}

const anyOfRequestsSucceeded = (calls: ICallWithRequest[], expected: unknown): boolean => {
  for (const call of calls) {
    if (equal(call[0].request, expected)) {
      return true
    }
  }
  return false
}

const equal = (request: Record<string, unknown>, expected: unknown): boolean => {
  try {
    expect(request).toEqual(expected)
    return true
  } catch (e) {
    return false
  }
}
