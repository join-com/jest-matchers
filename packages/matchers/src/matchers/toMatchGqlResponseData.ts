import { diff, matcherHint, printExpected, printReceived } from 'jest-matcher-utils'

export interface IGqlResponse {
  body: {
    singleResult: {
      data?: unknown
    }
  }
}

export const toMatchGqlResponseData = <E extends object>(
  response: IGqlResponse,
  expectedData: E,
): jest.CustomMatcherResult => {
  const responseData = response.body.singleResult.data
  const pass = responseMatches(responseData, expectedData)
  return {
    pass,
    message: () => {
      const hintString = matcherHint('.toMatchGqlResponseData')
      const expectedString = printExpected(expectedData)
      const mismatchString = formatMismatched(responseData, expectedData)
      return `${hintString}\n\n` + `Expected response data to match:\n${expectedString}\n` + mismatchString
    },
  }
}

const formatMismatched = (responseData: unknown, expectedData: unknown): string => {
  const diffString = diff(responseData, expectedData) ?? ''

  const receivedString = printReceived(responseData)
  const message = `${receivedString}\n` + 'Difference:\n' + `${diffString}`

  return message
}

const responseMatches = <E extends object>(responseData: unknown, expectedData: E): boolean => {
  try {
    expect(responseData).toMatchObject(expectedData)
    return true
  } catch (_e) {
    return false
  }
}
