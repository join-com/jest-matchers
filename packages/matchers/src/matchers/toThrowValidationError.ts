import equal from 'fast-deep-equal'
import { printDiffOrStringify } from 'jest-matcher-utils'
import { ValidationError } from '../ValidationError'

function getThrown(received: Error | (() => void)): Error | undefined {
  if (received instanceof Error) {
    return received
  }

  try {
    received()
  } catch (error) {
    return error as Error
  }

  return
}

export function toThrowValidationError<T extends ValidationError>(
  received: Error | (() => void),
  expected: T,
): jest.CustomMatcherResult {
  const thrown = getThrown(received)
  if (!thrown) {
    return {
      pass: false,
      message: () => 'Expected to receive an error or a function that throws an error',
    }
  }

  const isExpectedClass = thrown instanceof expected.constructor
  if (!isExpectedClass) {
    return {
      pass: false,
      message: () =>
        `Expected error to be an instance of ${expected.constructor.name} but got ${thrown.constructor.name}`,
    }
  }

  const receivedFields = (thrown as ValidationError).fields
  const areFieldsMatching = equal(receivedFields, expected.fields)
  if (!areFieldsMatching) {
    return {
      pass: false,
      message: () => printDiffOrStringify(receivedFields, expected.fields, 'Expected fields', 'Received fields', true),
    }
  }

  return { pass: true, message: () => '' }
}
