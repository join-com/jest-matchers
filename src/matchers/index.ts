import { toHaveBeenRequestedWith } from './toHaveBeenRequestedWith';

// tslint:disable-next-line
const jestExpect = global['expect'];

if (jestExpect !== undefined) {
  jestExpect.extend({
    toHaveBeenRequestedWith,
  });
} else {
  /* tslint:disable:no-console */
  console.error(`
    Unable to find Jest's global expect.
    Please check you have added jest-matchers to setupFilesAfterEnv section of jest config
  `);
  /* tslint:enable:no-console */
}
