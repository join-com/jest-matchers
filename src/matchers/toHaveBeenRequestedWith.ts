import { equals } from 'expect/build/jasmine_utils';
import diff from 'jest-diff';
import { matcherHint, printExpected, printReceived } from 'jest-matcher-utils';

export const toHaveBeenRequestedWith = (
  { mock }: jest.Mock<any>,
  expected: object,
) => {
  const received = { ...mock.calls[0][0].request };
  const pass = equals(received, expected);
  return {
    pass,
    message: () => {
      const diffString = diff(expected, received);
      const hintString = matcherHint('.toHaveBeenRequestedWith');
      const expectedString = printExpected(expected);
      const receivedString = printReceived(received);
      return (
        `${hintString}\n\n` +
        `Expected value to equal:\n${expectedString}\n` +
        `Received:\n${receivedString}\n` +
        `Difference:\n${diffString}`
      );
    },
  };
};
