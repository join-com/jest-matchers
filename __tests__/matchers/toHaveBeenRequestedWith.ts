import { toHaveBeenRequestedWith } from '../../src/matchers/toHaveBeenRequestedWith';

expect.extend({ toHaveBeenRequestedWith });

describe('toHaveBeenRequestedWith', () => {
  const mockFn = jest.fn();

  const request = {
    a: 1,
    b: 2,
  };

  beforeEach(() => {
    mockFn.mockClear();
  });

  it('passes when mock was called with request', () => {
    mockFn({ request });

    expect(mockFn).toHaveBeenRequestedWith({
      a: 1,
      b: 2,
    });
  });

  it('fails when mock was not called', () => {
    const validate = () => {
      expect(mockFn).toHaveBeenRequestedWith({
        a: 1,
        b: 2,
      });
    };

    expect(validate).toThrowErrorMatchingSnapshot();
  });

  it('fails when mock was called with different values', () => {
    mockFn({ request });

    const validate = () => {
      expect(mockFn).toHaveBeenRequestedWith({
        a: 1,
        b: 1,
      });
    };

    expect(validate).toThrowErrorMatchingSnapshot();
  });

  it('passes when one of three calls was correct', () => {
    const falseRequest = {
      a: 1,
      b: 1,
    };
    mockFn({ request: falseRequest });
    mockFn({ request });
    mockFn({ request: falseRequest });

    expect(mockFn).toHaveBeenRequestedWith({
      a: 1,
      b: 2,
    });
  });

  it('fails when all calls were incorrect', () => {
    const falseRequest = {
      a: 1,
      b: 1,
    };
    mockFn({ request: falseRequest });
    mockFn({ request: falseRequest });
    mockFn({ request: falseRequest });

    const validate = () => {
      expect(mockFn).toHaveBeenRequestedWith({
        a: 1,
        b: 2,
      });
    };

    expect(validate).toThrowErrorMatchingSnapshot();
  });
});
