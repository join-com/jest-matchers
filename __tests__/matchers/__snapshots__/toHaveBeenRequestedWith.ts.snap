// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toHaveBeenRequestedWith fails when mock was called with different values 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).toHaveBeenRequestedWith([22m[32mexpected[39m[2m)[22m

Expected value to equal:
[32m{\\"a\\": 1, \\"b\\": 1}[39m
Received:
[31m{\\"a\\": 1, \\"b\\": 2}[39m
Difference:
[32m- Expected[39m
[31m+ Received[39m

[2m  Object {[22m
[2m    \\"a\\": 1,[22m
[32m-   \\"b\\": 1,[39m
[31m+   \\"b\\": 2,[39m
[2m  }[22m"
`;

exports[`toHaveBeenRequestedWith fails when mock was not called 1`] = `"Cannot read property '0' of undefined"`;
